src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,320 @@
+    stagnation_threshold
+        See `stagnation_window`.
+    error_aware
+        If `True`, perform an additional error aware check during the line search phase.
+        Intended for use with the trust region algorithm. Requires the parameters `beta`
+        and `radius` to be set.
+    error_criterion
+        The additional error criterion used to check model confidence in the line search.
+    beta
+        Intended for use with the trust region algorithm. Indicates the factor for checking
+        if the current value is close to the trust region boundary. See `error_aware`.
+    radius
+        Intended for use with the trust region algorithm. Indicates the radius of the
+        current trust region. See `error_aware`.
+
+    Returns
+    -------
+    mu
+        |Numpy array| containing the computed parameter.
+    data
+        Dict containing the following fields:
+
+            :mus:                       `list` of parameters after each iteration.
+            :foc_norms:                 |NumPy array| of the first order criticity norms
+                                        after each iteration.
+            :update_norms:              |NumPy array| of the norms of the update vectors
+                                        after each iteration.
+            :iterations:                Number of total BFGS iterations.
+            :line_search_iterations:    |NumPy array| of the number of line search
+                                        iterations per BFGS iteration.
+
+    Raises
+    ------
+    BFGSError
+        Raised if the BFGS algorithm failed to converge.
+    """
+    logger = getLogger('pymor.algorithms.bfgs')
+
+    data = {}
+
+    assert model.output_functional is not None
+    output = lambda m: model.output(m)[0, 0]
+
+    if parameter_space is None:
+        parameter_space = model.parameters.space(-np.inf, np.inf)
+
+    if initial_guess is None:
+        initial_guess = parameter_space.sample_randomly(1)[0]
+        mu = initial_guess.to_numpy()
+    else:
+        mu = initial_guess.to_numpy() if isinstance(initial_guess, Mu) else initial_guess
+
+    if error_aware:
+        assert error_criterion is not None and beta is not None and radius is not None
+        assert callable(error_criterion)
+
+    gradient = model.parameters.parse(model.output_d_mu(mu)).to_numpy()
+    current_output = output(mu)
+    eps = np.linalg.norm(gradient)
+
+    # compute norms
+    mu_norm = np.linalg.norm(mu)
+    update_norms = []
+    foc_norms = []
+    line_search_iterations = []
+    data['mus'] = [mu.copy()]
+
+    hessian = np.eye(mu.size)
+
+    first_order_criticity = output_diff = mu_diff = update_norm = 1e6
+    iteration = 0
+    while True:
+        if iteration >= miniter:
+            if output_diff < rtol:
+                logger.info(f'Relative tolerance of {rtol} for output difference reached. Converged.')
+                break
+            if mu_diff < rtol:
```

Comment:
I don't think that two different parameters will matter here if the tolerances are both chosen small enough. On the other hand this would be pretty easy to add, so we might as well, right?

Reply:
Two tolerances certainly don't hurt and increases user-input flexibility.